---
title: "scRNA-seq_lipidomics_integration_01"
author: "Kari Neier"
date: "11/15/2021"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setting up

Loading libraries, setting working directory, set up parallelization

```{r setting up}

setwd("/Users/karineier/Documents/scRNA-seq")

packages <- c("edgeR", "tidyverse", "RColorBrewer", "org.Mm.eg.db", "AnnotationDbi", "EnhancedVolcano", "enrichR", "openxlsx", "glue", "Glimma", "DMRichR", "magrittr", "variancePartition", "UpSetR", "ComplexUpset")
stopifnot(suppressMessages(sapply(packages, require, character.only=TRUE)))

enrichR:::.onAttach()

param <- SnowParam(2, "SOCK", progressbar = TRUE)
register(param)

```

## Loading in Data Organized for Differential Expression

- DGEList = List of DGEList objects 
- design = design matrix

``` {r loading data}

load("scRNA-seq_diffExp_0.RData")

```

## Quality Control and Filtering

Currently filtering out genes if they are not expressed in at least 50% of the cells with at least one read. This results in filtering out the vast majority of genes. 

### log-CPM density plots

```{r log-CPM and filtering}

for(i in cell_types) {
  
  dir.create(glue::glue("{i}"))
  dir.create(glue::glue("{i}/QC"))
  dir.create(glue::glue("{i}/plotData"))
  
  L <- mean(DGEList[[i]]$samples$lib.size)*1e-6
  M <- median(DGEList[[i]]$samples$lib.size)*1e-6
  
  logCPM <- cpm(DGEList[[i]], log = TRUE)
  #logCPM.cutoff <- log2(1/M + 1/L)
  nsamples <- ncol(DGEList[[i]])
  col <- brewer.pal(nsamples, "Paired")
  
  pdf(glue::glue("{i}/QC/density_plot.pdf"), height=8.5, width=11)
  par(mfrow=c(1,2))
  
  plot(density(logCPM[,1]), col=col[1], lwd=2, las=2, main="", xlab="")
  title(main="A. Raw data", xlab="Log-CPM")
  #abline(v=logCPM.cutoff, lty=3)
  for(j in 2:nsamples){
    den <- density(logCPM[,j])
    lines(den$x, den$y, col=col[j], lwd=2)
  }
  #legend("topright", design$cell_ID, text.col=col, bty="n", cex=0.5)
    
    rawCount <- dim(DGEList[[i]])

# Filtering
    keep <- rowSums(DGEList[[i]]$counts==0, na.rm=TRUE) < 
      0.50*ncol(DGEList[[i]]$counts) # filtering genes where less than 50% of cells express the gene
    
    DGEList[[i]] <- DGEList[[i]][keep,, keep.lib.sizes = FALSE] %>%
      calcNormFactors()
    
    filterCount <- dim(DGEList[[i]])
    
    print(glue::glue("{100 - round((filterCount[1]/rawCount[1])*100)}% of genes were filtered from {rawCount[2]} cells, \\ where there were {rawCount[1]} genes before filtering and {filterCount[1]} genes after filtering for {i}"))
    
    # Filtered density plot of log-CPM values
    logCPM <- cpm(DGEList[[i]], log=TRUE)
    plot(density(logCPM[,1]), col=col[1], lwd = 2, las=2, main="", xlab="")
    title(main = "B. Filtered data", xlab="Log-cpm")
    #abline(v=logCPM.cutoff, lty=3)
    for(j in 2:nsamples){
      den <- density(logCPM[,j])
      lines(den$x, den$y, col = col[j], lwd=2)
      }
    dev.off()

}

```

## Sample normalization, Differential Expression analysis, and Pathway analysis

Normalizing samples with limma voom with dream weights using the variancePartition package. This allows us to model mouse ID as a random effect. 

Differential expression analysis is carried out for each of the top differentially abundant brain lipids in mutant vs. wild-type (p<0.05). The resulting differential expressed genes (DEGs) represent genes whose expression is significantly associated with that particular brain lipid in a particular cell type. Thus, you will have a list of DEGs for each lipid and for each cell type. 

Pathway analysis is performed using enrichR and then slimming similar GO terms with DMRichR.



``` {r normalizing samples, DEG analysis, Pathway analysis}

for (i in cell_types) {
  
  print(glue::glue("Normalizing {i} cells"))
  
  design.new <- design %>%
    dplyr::filter(cell_type == i)
  
  colnames(design.new) = gsub(" ", "_", colnames(design.new))
  colnames(design.new) = gsub(";_$", "", colnames(design.new))
  colnames(design.new) = gsub("\\:", "_", colnames(design.new))
  colnames(design.new) = gsub("\\;", "", colnames(design.new))
  colnames(design.new) = gsub("\\/", "_", colnames(design.new))
  
  samples.idx <- pmatch(design.new$cell_ID, rownames(DGEList[[i]]$samples))
  design.new <- design.new[order(samples.idx),]
  stopifnot(rownames(DGEList[[i]]$samples)==design.new$cell_ID)
  
  brain_lipids_sig = colnames(design.new[,c(7:ncol(design))])
  
  design.new[brain_lipids_sig] <- sapply(design.new[brain_lipids_sig], as.numeric)
  sapply(design.new, class)
  
  for (j in brain_lipids_sig) {
    
    model <- as.formula(glue::glue("~ {j} + (1|sample_ID)"))
    
    # Voom with dream weights 
    
    pdf(glue::glue("{i}/QC/voom_mean-variance_trend_{j}.pdf"), height=8.5, width=11)
    voomLogCPM <- variancePartition::voomWithDreamWeights(DGEList[[i]],
                                                          model,
                                                          design.new,
                                                          plot=TRUE)
    dev.off()
    
    # Boxplots of logCPM values before and after voom normalization
    pdf(glue::glue("{i}/QC/normalization_boxplots_{j}.pdf"), height=8.5, width=11)
    par(mfrow=c(1,2))
    
    logCPM <- cpm(DGEList[[i]], log=TRUE)
    col <- brewer.pal(nsamples, "Paired")
    boxplot(logCPM, las = 2, col = col, main = "")
    title(main="A. Unnormalized data", ylab="Log-cpm")
    
    boxplot(voomLogCPM$E, las = 2, col = col, main="")
    title(main="B. Normalized data", ylab="Log-cpm")
    
    dev.off()
    
    # Fitting linear models in limma
    
    print(glue::glue("Testing {i} cells for differential expression of {j}"))
    
    fit <- variancePartition::dream(voomLogCPM,
                                    model,
                                    design.new)
    
    head(coef(fit))
    
    # Save normalized expression values
    voomLogCPM$E %>%
      as.data.frame() %>%
      tibble::rownames_to_column(var = "geneSymbol") %>%
      openxlsx::write.xlsx(glue::glue("{i}/plotData/voomLogCPM_{j}.xlsx"))
    
    # Create DEG List 
    
    print(glue::glue("Creating DEG list of {i} cells for {j}"))
    
    dfit <- fit %>%
      contrasts.fit(coef=glue::glue("{j}"))
    
    pdf(glue::glue("{i}/QC/final_model_mean-variance_trend_{j}.pdf"), height=8.5, width=11)
    
    plotSA(dfit, main=glue::glue("Final model {i} {j}: Mean-variance trend"))
    
    dev.off()
    
    # Top DEGs
    
    DEGs <- dfit %>%
      topTable(sort.by = "P", n = Inf) %>%
      rownames_to_column() %>%
      tibble::as_tibble() %>%
      dplyr::rename(SYMBOL = rowname) %>%
      dplyr::mutate(FC = dplyr::case_when(logFC >0 ~ 2^logFC,
                                          logFC <0 ~ -1/(2^logFC))) %>%
      dplyr::select(SYMBOL, FC, logFC, P.Value, adj.P.Val, AveExpr, t, z.std) %T>%
      openxlsx::write.xlsx(file=glue::glue("{i}/DEGs_{j}.xlsx")) %>%
      dplyr::filter(P.Value < 0.05) %T>%
      openxlsx::write.xlsx(file=glue::glue("{i}/sig_DEGs_{j}.xlsx"))
    
    print(glue::glue("GO and Pathway analysis of {i} cells and {j}"))
    
    tryCatch({
      DEGs %>% 
        dplyr::select(SYMBOL) %>%
        purrr::flatten() %>%
        enrichR::enrichr(c("GO_Biological_Process_2018",
                           "GO_Molecular_Function_2018",
                           "GO_Cellular_Component_2018",
                           "KEGG_2019_Mouse",
                           "Panther_2016",
                           "Reactome_2016",
                           "RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO")) %>%
        purrr::set_names(names(.) %>% stringr::str_trunc(31, ellipsis="")) %T>%
        openxlsx::write.xlsx(file=glue::glue("{i}/enrichr_{j}.xlsx")) %>%
        DMRichR::slimGO(tool = "enrichR",
                        annoDb = "org.Mm.eg.db",
                        plots = FALSE) %T>%
        openxlsx::write.xlsx(file = glue::glue("{i}/rrvgo_enrichr_{j}.xlsx")) %>%
        DMRichR::GOplot() %>%
        ggplot2::ggsave(glue::glue("{i}/enrichr_plot_{j}.pdf"),
                        plot = .,
                        device = NULL,
                        height = 8.5,
                        width = 10) },
      error = function(error_condition) {
        print(glue::glue("ERROR: Gene Ontology pipe did not finish for {i} cells and {j}"))
      })
    print(glue::glue("The pipeline has finished for {i} cells and {j}"))
  }
}

```


```


