---
title: "Integration of scRNA-seq and Lipidomics Data - Mecp2 Project"
author: "Kari Neier"
date: "9/14/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

## Background

This experiment is designed to characterize transcriptomic profiles of Mecp2-e1 mutant vs. wild-type mice in the brain at the single-cell level. In addition, we also have lipidomic data from brain tissue of these same mice, and are interested in determining the following:

1) Which cell types have the strongest associations between gene expression and lipids that were differentially abundant in Mecp2-e1 mutants vs. wild-types?
2) Which genes are associated with brain lipids in diferent cell types? Are these genes also differentially expressed in Mecp2-e1 mutant vs. wild-types?

### Approach

Use limma to evaluate associations between lipids and gene expression, stratified by cell type. 

### The Data

- Labeled Lipids from an untargeted lipidomic analysis of brain tissue from male Mecp2-e1 mutants and wild-type littermate controls
- scRNA-seq data generated by Osman with data processed and cells labeled 

## Setting up

Loading packages, loading in the Seurat Object that was generated by Osman, and loading brain lipidomics data

```{r setting_up}
library(Seurat)
library(limma)
library(edgeR)
library(ggplot2)
library(tidyr)

load("/Users/karineier/Documents/Mecp2/scRNA-seq/rett_P120_with_labels_proportions3.rda")

brain_lipids = read.csv("/Users/karineier/Documents/Mecp2/Brain_Lipidomics/Males_Brain_semiquant_data_normalized.csv")

brain_lipids_stats = read.csv("/Users/karineier/Documents/Mecp2/Brain_Lipidomics/Males_Brain_Lipidomics_Semiquant_t_test_annotated.csv") # contains p-values and FDRs for brain lipids in males

```

## Formatting lipidomic data

Reorganizing lipidomic data for use in limma and filtering out lipids with p-value >0.05. Also removing data from animals without scRNA-seq data available.

```{r lipidomic_data_formatting}
brain_lipids.0 = as.data.frame(t(brain_lipids[-1,-1]))
colnames(brain_lipids.0) = gsub("\\[M.*", "", brain_lipids$X[-1])
rownames(brain_lipids.0) = gsub("X", "19-0", rownames(brain_lipids.0))

brain_lipids_stats$Annotation = gsub("\\[M.*", "", brain_lipids_stats$Annotation)

brain_lipids_sig = brain_lipids_stats$Annotation[which(brain_lipids_stats$p.value<0.05)]

brain_lipids.1 = brain_lipids.0[,brain_lipids_sig]

scRNAseq_animal_IDs = c("19-0112", "19-0113", "19-0114", "19-0117")

brain_lipids.2 = brain_lipids.1[scRNAseq_animal_IDs,]

head(brain_lipids.2)

```

## Extracting RNA-seq count data and design data

The portions of the Seurat object containing relevant data that will be used/potentially be used in differential expression analyses are:

1) @assays$RNA@counts = gene expression data for each cell (16790 in this case)
2) @meta.data$celltype.call = a cell type identifier for each cell 
3) @meta.data$orig.ident = the animal ID for each cell
4) @meta.data$batchid = batch ID
5) @meta.data$cell.cycle = cell cycle

```{r extracting_data_from_Seurat}

expr_matrix = experiment.aggregate@assays$RNA@counts

design = data.frame(cell_type = experiment.aggregate@meta.data$celltype.call, 
                    sample_ID = experiment.aggregate@meta.data$orig.ident,
                    cell_ID = colnames(expr_matrix),
                    batch = experiment.aggregate@meta.data$batchid,
                    cell_cycle = experiment.aggregate@meta.data$cell.cycle,
                    genotype = ifelse(grepl("WT", experiment.aggregate@meta.data$orig.ident)=="TRUE", "WT", "MUTANT"))

design = design %>%
  dplyr::mutate_if(is.character, as.factor)
  

head(design)
head(expr_matrix[1:5, 1:5])
```
## Formatting gene expression data for limma

Subsetting gene expression matrices by cell type and generating a list of expression matrices called expr_matrix_list

``` {r formatting_data}

cell_types = as.factor(levels(design$cell_type))
expr_matrix_list = vector(mode="list", length=nlevels(design$cell_type))

expr_matrix_list = lapply(levels(cell_types), function(x) {
  expr_matrix[,design$cell_ID[which(design$cell_type==x)]]
})

names(expr_matrix_list) = as.character(cell_types)

```

## Adding lipidomic data to the design matrix

Combining lipidomic data to the design matrix 

``` {r add_lipidomics_to_design}

brain_lipids.2$sample_ID = levels(design$sample_ID)
design = merge(design, brain_lipids.2, by="sample_ID")

```

## Creating DGEList 
DGEList for each cell type, as a list. Each element of the list is a DGEList for each cell type.

``` {r creating DGEList}

DGEList = vector(mode="list", length=nlevels(design$cell_type))

DGEList = lapply(cell_types, function(x) {
  DGEList(expr_matrix_list[[x]])
})

names(DGEList) = cell_types

```

## Saving DGEList for Differential Expression Analysis 
Saving the DGEList as an R object for downstream differential expression analysis

``` {r saving DGEList}

save(DGEList, design, cell_types, file="scRNA-seq_diffExp_0.R")


```


``` {r exploring DGEList and design matrix}

DGEList$Astro$samples$lib.size
DGEList[[1]]$samples$lib.size

```


