---
title: "scRNA-seq_lipidomics_integration_01"
author: "Kari Neier"
date: "11/15/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="/Users/karineier/Documents/scRNA-seq/Females")
```

## Setting up

Loading libraries, setting working directory, set up parallelization

```{r setting up}

packages <- c("edgeR", "tidyverse", "RColorBrewer", "org.Mm.eg.db", "AnnotationDbi", "EnhancedVolcano", "enrichR", "openxlsx", "glue", "Glimma", "DMRichR", "magrittr", "variancePartition", "UpSetR", "ComplexUpset")
stopifnot(suppressMessages(sapply(packages, require, character.only=TRUE)))

enrichR:::.onAttach()

param <- SnowParam(2, "SOCK", progressbar = TRUE)
register(param)

```

## Loading in Data Organized for Differential Expression

- DGEList = List of DGEList objects 
- design = design matrix

``` {r loading data}

load("scRNA-seq_diffExp_0.R")

```

## Quality Control and Filtering

Note: May need to explore additional filtering methods

### Removing cell types with too-few cells

``` {r subsetting_DGEList}

DGEList = DGEList[-c(2,10)] # Removing Endo and Peri cells; Endo = 1 cell and Peri = 3 cells

cell_types = cell_types[-c(2,10)]

```


### log-CPM density plots

```{r log-CPM and filtering}

for(i in cell_types) {
  
  dir.create(glue::glue("{i}"))
  dir.create(glue::glue("{i}/QC"))
  dir.create(glue::glue("{i}/plotData"))
  
  L <- mean(DGEList[[i]]$samples$lib.size)*1e-6
  M <- median(DGEList[[i]]$samples$lib.size)*1e-6
  
  logCPM <- cpm(DGEList[[i]], log = TRUE)
  #logCPM.cutoff <- log2(1/M + 1/L)
  nsamples <- ncol(DGEList[[i]])
  col <- brewer.pal(nsamples, "Paired")
  
  pdf(glue::glue("{i}/QC/density_plot.pdf"), height=8.5, width=11)
  par(mfrow=c(1,2))
  
  plot(density(logCPM[,1]), col=col[1], lwd=2, las=2, main="", xlab="")
  title(main="A. Raw data", xlab="Log-CPM")
  #abline(v=logCPM.cutoff, lty=3)
  for(j in 2:nsamples){
    den <- density(logCPM[,j])
    lines(den$x, den$y, col=col[j], lwd=2)
  }
  #legend("topright", design$cell_ID, text.col=col, bty="n", cex=0.5)
    
    rawCount <- dim(DGEList[[i]])

# Filtering
    keep <- rowSums(DGEList[[i]]$counts==0, na.rm=TRUE) < 
      0.50*ncol(DGEList[[i]]$counts) # filtering genes where less than 50% of cells express the gene
    
    DGEList[[i]] <- DGEList[[i]][keep,, keep.lib.sizes = FALSE] %>%
      calcNormFactors()
    
    filterCount <- dim(DGEList[[i]])
    
    print(glue::glue("{100 - round((filterCount[1]/rawCount[1])*100)}% of genes were filtered from {rawCount[2]} cells, \\ where there were {rawCount[1]} genes before filtering and {filterCount[1]} genes after filtering for {i}"))
    
    # Filtered density plot of log-CPM values
    logCPM <- cpm(DGEList[[i]], log=TRUE)
    plot(density(logCPM[,1]), col=col[1], lwd = 2, las=2, main="", xlab="")
    title(main = "B. Filtered data", xlab="Log-cpm")
    #abline(v=logCPM.cutoff, lty=3)
    for(j in 2:nsamples){
      den <- density(logCPM[,j])
      lines(den$x, den$y, col = col[j], lwd=2)
      }
    dev.off()

}

```

## Saving filtered data

``` {r saving_filtered_data}

save(DGEList, design, cell_types, file="Filtered_female_data.RData")

``` 

